walkeeper/src/replication.rs

```diff
@@ -194,8 +194,14 @@ impl ReplicationConn {
                 break;
             }
         }
-        let (_, timeline) = swh.timeline.get().get_end_of_wal();
-        info!("Start replication from {}", start_pos);
+        let (wal_end, timeline) = swh.timeline.get().get_end_of_wal();
+        let stop_pos: Lsn;
+        if swh.appname == Some("wal_proposer_recovery".to_string()) {
```

Comment:
I'd elaborate here, e.g.:

```
// wal_proposer_recovery gets special handling because it is:
// a) part of the consensus protocol,
// b) that has to synchronize Safekeepers with each other,
// c) and needs all flushed WAL from all Safekeepers to do so even if it's not committed.
// So, we serve it all existing WAL (even uncommitted) and exit immediately.
// We cannot continue replication because WAL may be overwritten after proposer recovers and overwrites uncommitted history.
// If we receive new WAL during this replication, that WAL should originate with another proposer node. Hence, TODO TODO TODO
```

Is this accurate? I'm actually not sure why it's ok to set `stop_pos = wal_end` here and not care about sending updated WAL. Say, what if we had no uncommitted changes, but then they suddently appear from another proposer node... Not sure.

Reply:
I've added something along these lines. Minor correction: 'We have to exit otherwise' -- no, walproposer actually knows when it got enough WAL and will end streaming on its own. We just need to have some border till which safekeeper has WAL available, and stop_pos serves this purpose instead of `wait_for_lsn` used normally.